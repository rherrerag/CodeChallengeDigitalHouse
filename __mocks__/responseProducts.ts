const responseOK = {
  data: [
    {
      createdAt: '2023-03-11T04:50:50.205Z',
      product: 'Recycled Plastic Tuna',
      points: 92984,
      image: 'https://loremflickr.com/640/480/technics',
      is_redemption: false,
      id: '2',
    },
    {
      createdAt: '2023-03-11T04:35:35.259Z',
      product: 'Fantastic Granite Bacon',
      points: 42416,
      image: 'https://loremflickr.com/640/480/technics',
      is_redemption: false,
      id: '3',
    },
    {
      createdAt: '2023-03-11T13:57:33.428Z',
      product: 'Fantastic Fresh Gloves',
      points: 23913,
      image: 'https://loremflickr.com/640/480/city',
      is_redemption: true,
      id: '4',
    },
    {
      createdAt: '2023-03-11T12:17:47.118Z',
      product: 'Rustic Rubber Bacon',
      points: 69814,
      image: 'https://loremflickr.com/640/480/people',
      is_redemption: true,
      id: '5',
    },
    {
      createdAt: '2023-02-04T07:48:16.249Z',
      product: 'Licensed Metal Salad',
      points: 93367,
      image: 'https://loremflickr.com/640/480/food',
      is_redemption: false,
      id: '12',
    },
  ],
  error: false,
  isCancel: false,
};

const responseError = {
  data: {message: 'Error', code: 'ERR'},
  error: true,
};

export {responseOK, responseError};
