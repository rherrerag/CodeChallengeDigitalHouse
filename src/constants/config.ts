const config = {
  FILTER: {
    MONTH: '3',
    YEAR: '2023',
  },
};

export default config;
