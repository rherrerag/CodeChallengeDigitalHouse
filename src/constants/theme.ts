import {extendTheme} from 'native-base';

const customTheme = extendTheme({
  colors: {
    primary: {
      100: '#334FFA',
      200: '#CFD6FF',
    },
    text: {
      title: '#020202',
      subtitle: '#9B9898',
      normal: '#000000',
    },
  },
});

export {customTheme};
