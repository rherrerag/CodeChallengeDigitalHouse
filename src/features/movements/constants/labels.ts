const labels = {
  WELCOME: {
    MESSAGE: 'Bienvenido de Vuelta!',
    NAME: 'Ruben Rodriguez',
  },
  SUBTITLES: {
    POINTS: 'TUS PUNTOS',
    MOVEMENTS: 'TUS MOVIMIENTOS',
  },
  CARD: {
    POINTS: 'pts',
  },
  BUTTONS: {
    EARNED: 'Ganados',
    REDEEMED: 'Canjeados',
    ALL: 'Todos',
  },
  LOADING: {
    TITLE: 'Cargando',
  },
};

export default labels;
