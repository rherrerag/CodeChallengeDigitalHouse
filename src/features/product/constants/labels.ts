const labels = {
  PRODUCT_DETAIL: {
    SUBTITLE_DETAIL: 'Detalles del producto:',
    SUBTITLE_SUMMARY: 'Con esta compra acumulaste:',
    POINTS_TEXT: 'puntos',
    DATE_SUBSTRING: 'Comprado el',
  },
  ACCEPT_BUTTON: {
    TITLE: 'Aceptar',
  },
};

export default labels;
