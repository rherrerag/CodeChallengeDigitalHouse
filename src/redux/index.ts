import {configureStore, ThunkAction, Action} from '@reduxjs/toolkit';
import {movementsReducer} from '../features/movements/reducers/movements.reducer';

export const store = configureStore({
  reducer: {
    movements: movementsReducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
