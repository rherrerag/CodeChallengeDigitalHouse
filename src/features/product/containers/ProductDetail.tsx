import React from 'react';
import {Box, ScrollView} from 'native-base';
import {Movement} from '../../movements/interfaces/movement.interface';
import {useRoute} from '@react-navigation/native';
import {useNavigation, RouteProp} from '@react-navigation/native';
import {StackNavigationProp} from '@react-navigation/stack';
import {SafeAreaView} from 'react-native-safe-area-context';
import {monthNames} from '../../../common/utils/dateUtils';
import {
  ProductDetailContainer,
  MainContainer,
  ProductImage,
  AcceptButton,
} from './ProductDetail.styled';
import labels from '../constants/labels';
import ProductTitle from '../components/ProductTitle';
import ProductInfo from '../components/ProductInfo';

export type RootStackParamList = {
  Movements: {item: Movement | undefined};
};

type MovementsScreenRouteProp = RouteProp<RootStackParamList, 'Movements'>;

const ProductDetail = () => {
  const route = useRoute<MovementsScreenRouteProp>();
  const {item} = route.params;
  const movement = item;
  const navigation = useNavigation<StackNavigationProp<RootStackParamList>>();
  let date = new Date(movement?.createdAt);
  let dateFormat: string = `${
    labels.PRODUCT_DETAIL.DATE_SUBSTRING
  } ${date.getDay()} de ${monthNames[date.getMonth()]}, ${date.getFullYear()}`;

  const goToMovements = () => navigation.navigate('Movements');
  const pointSummaryString =
    movement?.points.toLocaleString('es-MX') +
    ' ' +
    labels.PRODUCT_DETAIL.POINTS_TEXT;

  return (
    <MainContainer flex="1">
      <ProductTitle title={movement?.product} />
      <SafeAreaView>
        <ScrollView>
          <ProductDetailContainer>
            <Box shadow="9">
              <ProductImage source={{uri: movement?.image}} />
            </Box>
            <ProductInfo
              productDetailTitle={labels.PRODUCT_DETAIL.SUBTITLE_DETAIL}
              pointsSummaryTitle={labels.PRODUCT_DETAIL.SUBTITLE_SUMMARY}
              pointsSummary={pointSummaryString}
              dateFormatted={dateFormat}
            />
            <AcceptButton
              bgColor="primary.100"
              _text={{
                fontSize: 16,
                fontWeight: 800,
                color: 'white',
                lineHeight: 22,
              }}
              onPress={() => goToMovements()}>
              {labels.ACCEPT_BUTTON.TITLE}
            </AcceptButton>
          </ProductDetailContainer>
        </ScrollView>
      </SafeAreaView>
    </MainContainer>
  );
};

export default ProductDetail;
