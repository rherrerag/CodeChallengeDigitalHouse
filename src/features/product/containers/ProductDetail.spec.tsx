import React from 'react';
import {
  render,
  screen,
  fireEvent,
  waitFor,
  act,
} from '@testing-library/react-native';
import {responseOK} from '../../../../__mocks__/responseProducts';
import {Provider} from 'react-redux';
import ProductDetail from './ProductDetail';
import {store} from '../../../redux';
import mockFetch from 'jest-fetch-mock';
import {NativeBaseProvider} from 'native-base';
import {NavigationContainer} from '@react-navigation/native';
import {customTheme} from '../../../constants/theme';
import {Movement} from '../../movements/interfaces/movement.interface';

const movementMock: Movement = {
  createdAt: '2023-03-11T04:35:35.259Z',
  product: 'Fantastic Granite Bacon',
  points: 42416,
  image: 'https://loremflickr.com/640/480/technics',
  is_redemption: false,
  id: 3,
};

const inset = {
  frame: {x: 0, y: 0, width: 0, height: 0},
  insets: {top: 0, left: 0, right: 0, bottom: 0},
};

const ProductDetailFactory = async () => {
  mockFetch.mockResponse(req => {
    if (req.url.includes('products')) {
      return Promise.resolve(JSON.stringify(responseOK.data));
    }

    return Promise.reject(new Error('not mapped url'));
  });
  render(
    <Provider store={store}>
      <NativeBaseProvider theme={customTheme} initialWindowMetrics={inset}>
        <NavigationContainer>
          <ProductDetail />
        </NavigationContainer>
      </NativeBaseProvider>
    </Provider>,
  );
  await waitFor(() =>
    expect(screen.getAllByTestId('accept-button')).toBeTruthy(),
  );
};

const mockedDispatch = jest.fn();

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: jest.fn(),
      dispatch: mockedDispatch,
    }),
    useRoute: () => ({
      params: {
        item: movementMock,
      },
    }),
  };
});

describe('Product Detail Container', () => {
  beforeEach(() => {
    mockFetch.enableMocks();
    mockFetch.resetMocks();
    mockedDispatch.mockClear();
  });

  it('debería generar snapshot de container correctamente', async () => {
    await ProductDetailFactory();
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it('debería navegar a movimientos al presionar Aceptar', async () => {
    await ProductDetailFactory();
    await act(() => {
      fireEvent(screen.getByTestId('accept-button'), 'press');
    });
    expect(screen.findAllByText('TUS PUNTOS')).toBeTruthy();
  });
});
