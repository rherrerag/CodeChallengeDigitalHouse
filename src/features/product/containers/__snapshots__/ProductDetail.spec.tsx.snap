// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Detail Container deber√≠a generar snapshot de container correctamente 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    dataSet={{}}
    style={
      [
        {
          "flex": 1,
        },
        [
          {
            "backgroundColor": "white",
          },
          undefined,
        ],
      ]
    }
  >
    <View
      dataSet={{}}
      style={
        [
          {
            "backgroundColor": "#CFD6FF",
          },
          [
            {
              "paddingBottom": 24,
              "paddingLeft": 20,
              "paddingRight": 20,
              "paddingTop": 80,
            },
            undefined,
          ],
        ]
      }
    >
      <Text
        dataSet={{}}
        style={
          [
            {
              "backgroundColor": undefined,
              "color": "#171717",
              "fontFamily": undefined,
              "fontSize": 14,
              "fontStyle": "normal",
              "fontWeight": "400",
              "letterSpacing": 0,
              "lineHeight": 21,
              "textDecorationLine": undefined,
            },
            [
              {
                "fontSize": 24,
                "fontWeight": "800",
                "lineHeight": 24,
              },
              undefined,
            ],
          ]
        }
      >
        Fantastic Granite Bacon
      </Text>
    </View>
    <RNCSafeAreaView>
      <RCTScrollView
        contentContainerStyle={
          [
            {},
            {
              "dataSet": {},
            },
            {},
          ]
        }
        dataSet={{}}
        style={{}}
      >
        <View>
          <View
            dataSet={{}}
            style={
              [
                {},
                [
                  {
                    "backgroundColor": "#fff",
                    "paddingBottom": 150,
                    "paddingLeft": 20,
                    "paddingRight": 20,
                    "paddingTop": 20,
                  },
                  undefined,
                ],
              ]
            }
          >
            <View
              dataSet={{}}
              style={
                {
                  "elevation": 10,
                  "shadowColor": "#000000",
                  "shadowOffset": {
                    "height": 5,
                    "width": 0,
                  },
                  "shadowOpacity": 0.34,
                  "shadowRadius": 6.27,
                }
              }
            >
              <Image
                source={
                  {
                    "uri": "https://loremflickr.com/640/480/technics",
                  }
                }
                style={
                  [
                    {
                      "height": 340,
                      "width": "100%",
                    },
                    undefined,
                  ]
                }
              />
            </View>
            <Text
              dataSet={{}}
              style={
                [
                  {
                    "backgroundColor": undefined,
                    "color": "#9B9898",
                    "fontFamily": undefined,
                    "fontSize": 14,
                    "fontStyle": "normal",
                    "fontWeight": "400",
                    "letterSpacing": 0,
                    "lineHeight": 21,
                    "textDecorationLine": undefined,
                  },
                  [
                    {
                      "fontSize": 14,
                      "fontWeight": "800",
                      "lineHeight": 19,
                      "paddingTop": 32,
                    },
                    undefined,
                  ],
                ]
              }
            >
              Detalles del producto:
            </Text>
            <Text
              dataSet={{}}
              style={
                [
                  {
                    "backgroundColor": undefined,
                    "color": "#171717",
                    "fontFamily": undefined,
                    "fontSize": 14,
                    "fontStyle": "normal",
                    "fontWeight": "400",
                    "letterSpacing": 0,
                    "lineHeight": 21,
                    "textDecorationLine": undefined,
                  },
                  [
                    {
                      "fontSize": 16,
                      "fontWeight": "800",
                      "lineHeight": 22,
                      "paddingTop": 20,
                    },
                    undefined,
                  ],
                ]
              }
            >
              Comprado el 6 de Marzo, 2023
            </Text>
            <Text
              dataSet={{}}
              style={
                [
                  {
                    "backgroundColor": undefined,
                    "color": "#9B9898",
                    "fontFamily": undefined,
                    "fontSize": 14,
                    "fontStyle": "normal",
                    "fontWeight": "400",
                    "letterSpacing": 0,
                    "lineHeight": 21,
                    "textDecorationLine": undefined,
                  },
                  [
                    {
                      "fontSize": 14,
                      "fontWeight": "800",
                      "lineHeight": 19,
                      "paddingTop": 20,
                    },
                    undefined,
                  ],
                ]
              }
            >
              Con esta compra acumulaste:
            </Text>
            <Text
              dataSet={{}}
              style={
                [
                  {
                    "backgroundColor": undefined,
                    "color": "#171717",
                    "fontFamily": undefined,
                    "fontSize": 14,
                    "fontStyle": "normal",
                    "fontWeight": "400",
                    "letterSpacing": 0,
                    "lineHeight": 21,
                    "textDecorationLine": undefined,
                  },
                  [
                    {
                      "fontSize": 24,
                      "fontWeight": "800",
                      "lineHeight": 33,
                      "paddingBottom": 20,
                      "paddingTop": 20,
                    },
                    undefined,
                  ],
                ]
              }
            >
              42,416 puntos
            </Text>
            <View
              accessibilityRole="button"
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              colorScheme="primary"
              dataSet={{}}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                [
                  {
                    "alignItems": "center",
                    "backgroundColor": "#334FFA",
                    "borderRadius": 4,
                    "flexDirection": "row",
                    "justifyContent": "center",
                    "paddingBottom": 10,
                    "paddingLeft": 12,
                    "paddingRight": 12,
                    "paddingTop": 10,
                  },
                  [
                    {
                      "borderBottomLeftRadius": 10,
                      "borderBottomRightRadius": 10,
                      "borderTopLeftRadius": 10,
                      "borderTopRightRadius": 10,
                    },
                    undefined,
                  ],
                ]
              }
              testID="accept-button"
            >
              <View
                dataSet={{}}
                style={
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                  }
                }
                test={true}
              >
                <View
                  dataSet={{}}
                  style={{}}
                >
                  <Text
                    dataSet={{}}
                    style={
                      {
                        "backgroundColor": undefined,
                        "color": "#FFFFFF",
                        "fontFamily": undefined,
                        "fontSize": 16,
                        "fontStyle": "normal",
                        "fontWeight": "800",
                        "letterSpacing": 0,
                        "lineHeight": 22,
                        "textDecorationLine": undefined,
                      }
                    }
                  >
                    Aceptar
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </RCTScrollView>
    </RNCSafeAreaView>
  </View>
</RNCSafeAreaProvider>
`;
